File: app/src/test/java/com/esoxjem/movieguide/movies/sorting/SortingDialogPresenterTest.java
package com.esoxjem.movieguide.movies.sorting;

import com.esoxjem.movieguide.RxSchedulerRule;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * @author arunsasidharan
 */
@RunWith(MockitoJUnitRunner.class)
public class SortingDialogPresenterTest {
    @Rule
    public RxSchedulerRule rule = new RxSchedulerRule();
    @Mock
    private SortingContract.Interactor interactor;
    @Mock
    private SortingContract.View view;

    private SortingContract.Presenter presenter;

    @Before
    public void setUp() throws Exception {
        presenter = new SortingDialogPresenter(interactor);
        presenter.setView(view);
    }

    @Test
    public void shouldCheckPopularIfLastSavedOptionIsPopular() throws Exception {
        when(interactor.getSelectedSortingOption()).thenReturn(SortType.MOST_POPULAR.getValue());
        presenter.setLastSavedOption();
        verify(view).setPopularChecked();
    }

    @Test
    public void shouldCheckHighestRatedIfLastSavedOptionIsHighestRated() throws Exception {
        when(interactor.getSelectedSortingOption()).thenReturn(SortType.HIGHEST_RATED.getValue());
        presenter.setLastSavedOption();
        verify(view).setHighestRatedChecked();
    }

    @Test
    public void shouldCheckFavoritesIfLastSavedOptionIsFavorites() throws Exception {
        when(interactor.getSelectedSortingOption()).thenReturn(SortType.FAVORITES.getValue());
        presenter.setLastSavedOption();
        verify(view).setFavoritesChecked();
    }

    @Test
    public void shouldCheckNewestMoviesIfLastSavedOptionIsHighestRated() throws Exception {
        when(interactor.getSelectedSortingOption()).thenReturn(SortType.NEWEST.getValue());
        presenter.setLastSavedOption();
        verify(view).setNewestChecked();
    }


    @Test
    public void onPopularMoviesSelected() throws Exception {
        presenter.onPopularMoviesSelected();
        verify(interactor).setSortingOption(SortType.MOST_POPULAR);
    }

    @Test
    public void onHighestRatedMoviesSelected() throws Exception {
        presenter.onHighestRatedMoviesSelected();
        verify(interactor).setSortingOption(SortType.HIGHEST_RATED);
    }

    @Test
    public void onFavoritesSelected() throws Exception {
        presenter.onFavoritesSelected();
        verify(interactor).setSortingOption(SortType.FAVORITES);
    }

    @Test
    public void onNewestMoviesSelected() throws Exception {
        presenter.onNewestMoviesSelected();
        verify(interactor).setSortingOption(SortType.NEWEST);
    }


}

File: app/src/test/java/com/esoxjem/movieguide/movies/sorting/SortingDialogPresenterTest.kt
package com.esoxjem.movieguide.movies.sorting

import com.esoxjem.movieguide.RxSchedulerRule
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner

/**
 * @author arunsasidharan
 * @author nitishbhatt
 */
@RunWith(MockitoJUnitRunner::class)
class SortingDialogPresenterTest {
    @JvmField
    @Rule
    var rule = RxSchedulerRule()

    private val interactor: SortingContract.Interactor = mock()
    private val view: SortingContract.View = mock()
    private var presenter: SortingContract.Presenter = mock()

    @Before
    @Throws(Exception::class)
    fun setUp() {
        presenter = SortingDialogPresenter(interactor)
        presenter.setView(view)
    }

    @Test
    @Throws(Exception::class)
    fun shouldCheckPopularIfLastSavedOptionIsPopular() {
        //given
        whenever(interactor.getSelectedSortingOption()).thenReturn(SortType.MOST_POPULAR.value)
        //when
        presenter.setLastSavedOption()
        //then
        verify(view).setPopularChecked()
    }

    @Test
    @Throws(Exception::class)
    fun shouldCheckHighestRatedIfLastSavedOptionIsHighestRated() {
        //given
        whenever(interactor.getSelectedSortingOption()).thenReturn(SortType.HIGHEST_RATED.value)
        //when
        presenter.setLastSavedOption()
        //then
        verify(view).setHighestRatedChecked()
    }

    @Test
    @Throws(Exception::class)
    fun shouldCheckFavoritesIfLastSavedOptionIsFavorites() {
        //given
        whenever(interactor.getSelectedSortingOption()).thenReturn(SortType.FAVORITES.value)
        //when
        presenter.setLastSavedOption()
        //then
        verify(view).setFavoritesChecked()
    }

    @Test
    @Throws(Exception::class)
    fun shouldCheckNewestMoviesIfLastSavedOptionIsHighestRated() {
        //given
        whenever(interactor.getSelectedSortingOption()).thenReturn(SortType.NEWEST.value)
        //when
        presenter.setLastSavedOption()
        //then
        verify(view).setNewestChecked()
    }

    @Test
    @Throws(Exception::class)
    fun onPopularMoviesSelected() {
        //given

        //when
        presenter.onPopularMoviesSelected()
        //then
        verify(interactor).setSortingOption(SortType.MOST_POPULAR)
    }

    @Test
    @Throws(Exception::class)
    fun onHighestRatedMoviesSelected() {
        //given

        //when
        presenter.onHighestRatedMoviesSelected()
        //then
        verify(interactor).setSortingOption(SortType.HIGHEST_RATED)
    }

    @Test
    @Throws(Exception::class)
    fun onFavoritesSelected() {
        //given

        //when
        presenter.onFavoritesSelected()
        //then
        verify(interactor).setSortingOption(SortType.FAVORITES)
    }

    @Test
    @Throws(Exception::class)
    fun onNewestMoviesSelected() {
        //given

        //when
        presenter.onNewestMoviesSelected()
        //then
        verify(interactor).setSortingOption(SortType.NEWEST)
    }
}